src/lib/raft/raft.c

```diff
@@ -775,17 +843,20 @@ raft_sm_schedule_new_election_cb(struct ev_loop *loop, struct ev_timer *timer,
 	 */
 	assert(raft_is_fully_on_disk(raft));
 	raft_ev_timer_stop(loop, timer);
-	raft_sm_schedule_new_election(raft);
+	bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
+	if (is_changed)
+		raft_schedule_broadcast(raft);
+	else
+		raft_sm_election_update(raft);
```

Comment:
Why isn't it possible that both the change happened and need to update the election? What if this node was the last to see the leader and the map became zero just here?

**Update:** now I see, you actually call it always. But in case of a change it is called by `raft_notify_is_leader_seen()`. That is because the election update is not re-enterable. How about this:
```Diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index ccd716e22..f1a9090c0 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -231,9 +231,6 @@ raft_set_leader(struct raft *raft, uint32_t leader)
 	bool is_changed = raft_notify_is_leader_seen(raft, true, raft->self);
 	assert(is_changed);
 	(void)is_changed;
-	raft->leader_last_seen = raft_ev_now(raft_loop());
-	/* Leader is seen. Tell other nodes about its appearance. */
-	raft_schedule_broadcast(raft);
 }
 
 /**
@@ -246,9 +243,7 @@ raft_clear_leader(struct raft *raft)
 	if (raft->leader == raft->self)
 		raft->leader_last_seen = raft_ev_now(raft_loop());
 	raft->leader = 0;
-	bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
+	raft_notify_is_leader_seen(raft, false, raft->self);
 }
 
 /** If there is split vote, the node might reduce the next term delay. */
@@ -326,8 +321,23 @@ raft_sm_schedule_new_election(struct raft *raft);
 static inline void
 raft_sm_election_update(struct raft *raft)
 {
-	if (raft->is_candidate && raft->leader_witness_map == 0)
-		raft_sm_schedule_new_election(raft);
+	if (!raft->is_candidate)
+		return;
+	/*
+	 * Pre-vote protection. Every node must agree that the leader is gone.
+	 */
+	if (raft->leader_witness_map != 0)
+		return;
+	/*
+	 * Don't start anything new ever during a WAL write. Let it end and then
+	 * check for updates.
+	 */
+	if (raft->is_write_in_progress)
+		return;
+	/* Already have ongoing elections. */
+	if (raft->vote != 0)
+		return;
+	raft_sm_schedule_new_election(raft);
 }
 
 /**
@@ -600,7 +610,6 @@ raft_process_heartbeat(struct raft *raft, uint32_t source)
 	 */
 	if (raft->is_write_in_progress)
 		return;
-	raft->leader_last_seen = raft_ev_now(raft_loop());
 	/*
 	 * XXX: it may be expensive to reset the timer like that. It may be less
 	 * expensive to let the timer work, and remember last timestamp when
@@ -614,9 +623,7 @@ raft_process_heartbeat(struct raft *raft, uint32_t source)
 	} else {
 		assert(!raft_ev_timer_is_active(&raft->timer));
 	}
-	bool is_changed = raft_notify_is_leader_seen(raft, true, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
+	raft_notify_is_leader_seen(raft, true, raft->self);
 	raft_sm_wait_leader_dead(raft);
 }
 
@@ -843,11 +850,8 @@ raft_sm_election_update_cb(struct ev_loop *loop, struct ev_timer *timer,
 	 */
 	assert(raft_is_fully_on_disk(raft));
 	raft_ev_timer_stop(loop, timer);
-	bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
-	else
-		raft_sm_election_update(raft);
+	raft_notify_is_leader_seen(raft, false, raft->self);
+	raft_sm_election_update(raft);
 }
 
 static void
@@ -1068,16 +1072,21 @@ raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source)
 {
 	assert(source > 0 && source < VCLOCK_MAX);
-
 	bool was_seen;
 	if (is_leader_seen) {
 		was_seen = bit_set(&raft->leader_witness_map, source);
+		if (raft->self == source)
+			raft->leader_last_seen = raft_ev_now(raft_loop());
 	} else {
 		was_seen = bit_clear(&raft->leader_witness_map, source);
 		if (was_seen)
 			raft_sm_election_update(raft);
 	}
-	return was_seen != is_leader_seen;
+	if (was_seen == is_leader_seen)
+		return false;
+	if (raft->self == source)
+		raft_schedule_broadcast(raft);
+	return true;
 }
```
Now you can update the elections explicitly anywhere and not depend on whether it is called by some other function inside (I didn't test this particular diff though).

Reply:
I couldn't make `raft_sm_election_update` re-enterable. Mainly because it's hard to know whether it's already called or not.
Your check (`raft->vote != 0`) doesn't work because `raft->vote != 0` doesn't work because node has almost surely voted in the last term (for previous leader).

My attempt to fix it (`raft->vote != 0 && raft->leader == 0`) doesn't work either, because it's possible that a node resigns and then it may want to start elections in the same term later.

So, here's the diff. I've extracted `raft_leader_see`/`raft_leader_unsee` for simplicity.
But I couldn't fix the issue with `raft_sm_election_udpate`. Probably need to call it outside of `raft_leader_see/unsee` everywhere.

<details><summary>diff</summary>

```diff
diff --git a/src/lib/raft/raft.c b/src/lib/raft/raft.c
index 14557d905..2177cc2bc 100644
--- a/src/lib/raft/raft.c
+++ b/src/lib/raft/raft.c
@@ -218,26 +218,6 @@ raft_is_leader_seen(const struct raft *raft)
 static void
 raft_schedule_broadcast(struct raft *raft);
 
-/**
- * Clear leader-related fields on leader loss or resign and let everyone know
- * this node doesn't see the leader anymore.
- */
-static inline void
-raft_clear_leader(struct raft *raft)
-{
-	if (raft->leader == raft->self) {
-		/*
-		 * Update leader_last_seen when resigning so that leader_idle
-		 * starts counting from zero after the resign.
-		 */
-		raft->leader_last_seen = raft_ev_monotonic_now(raft_loop());
-	}
-	raft->leader = 0;
-	bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
-}
-
 /** If there is split vote, the node might reduce the next term delay. */
 static void
 raft_check_split_vote(struct raft *raft);
@@ -313,8 +293,15 @@ raft_sm_schedule_new_election(struct raft *raft);
 static inline void
 raft_sm_election_update(struct raft *raft)
 {
-	if (raft->is_candidate && raft->leader_witness_map == 0)
-		raft_sm_schedule_new_election(raft);
+	if (!raft->is_candidate)
+		return;
+	/*
+	 * Pre-vote protection. Every node must agree that the leader is gone.
+	 */
+	if (raft->leader_witness_map != 0)
+		return;
+
+	raft_sm_schedule_new_election(raft);
 }
 
 /**
@@ -421,6 +408,32 @@ raft_process_recovery(struct raft *raft, const struct raft_msg *req)
 	assert(!raft_is_enabled(raft));
 }
 
+/** Update raft state once leader is seen. */
+static bool
+raft_leader_see(struct raft *raft);
+
+/** Perform needed actions on leader loss. */
+static bool
+raft_leader_unsee(struct raft *raft);
+
+/**
+ * Clear leader-related fields on leader loss or resign and let everyone know
+ * this node doesn't see the leader anymore.
+ */
+static inline void
+raft_clear_leader(struct raft *raft)
+{
+	if (raft->leader == raft->self) {
+		/*
+		 * Update leader_last_seen when resigning so that leader_idle
+		 * starts counting from zero after the resign.
+		 */
+		raft->leader_last_seen = raft_ev_monotonic_now(raft_loop());
+	}
+	raft->leader = 0;
+	raft_leader_unsee(raft);
+}
+
 int
 raft_process_msg(struct raft *raft, const struct raft_msg *req, uint32_t source)
 {
@@ -563,8 +576,9 @@ raft_process_msg(struct raft *raft, const struct raft_msg *req, uint32_t source)
 	 * The message came from the leader itself. Can be sure it is visible to
 	 * this instance.
 	 */
-	if (raft_notify_is_leader_seen(raft, true, raft->self))
-		raft_schedule_broadcast(raft);
+	bool changed = raft_leader_see(raft);
+	assert(changed);
+	(void)changed;
 	/* New leader was elected. */
 	raft_sm_follow_leader(raft, source);
 	return 0;
@@ -608,9 +622,7 @@ raft_process_heartbeat(struct raft *raft, uint32_t source)
 	} else {
 		assert(!raft_ev_timer_is_active(&raft->timer));
 	}
-	bool is_changed = raft_notify_is_leader_seen(raft, true, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
+	raft_leader_see(raft);
 	raft_sm_wait_leader_dead(raft);
 }
 
@@ -733,7 +745,9 @@ raft_sm_become_leader(struct raft *raft)
 	assert(!raft->is_write_in_progress);
 	raft->state = RAFT_STATE_LEADER;
 	raft->leader = raft->self;
-	raft_notify_is_leader_seen(raft, true, raft->self);
+	bool changed = raft_leader_see(raft);
+	assert(changed);
+	(void)changed;
 	raft_ev_timer_stop(raft_loop(), &raft->timer);
 	/* State is visible and it is changed - broadcast. */
 	raft_schedule_broadcast(raft);
@@ -845,10 +859,8 @@ raft_sm_election_update_cb(struct ev_loop *loop, struct ev_timer *timer,
 	 */
 	assert(raft_is_fully_on_disk(raft));
 	raft_ev_timer_stop(loop, timer);
-	bool is_changed = raft_notify_is_leader_seen(raft, false, raft->self);
-	if (is_changed)
-		raft_schedule_broadcast(raft);
-	else
+	bool was_seen = raft_leader_unsee(raft);
+	if (!was_seen)
 		raft_sm_election_update(raft);
 }
 
@@ -1069,15 +1081,11 @@ bool
 raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 			   uint32_t source)
 {
-	assert(source > 0 && source < VCLOCK_MAX);
+	assert(source > 0 && source < VCLOCK_MAX && source != raft->self);
 
 	bool was_seen;
 	if (is_leader_seen) {
 		was_seen = bit_set(&raft->leader_witness_map, source);
-		if (source == raft->self) {
-			raft->leader_last_seen =
-				raft_ev_monotonic_now(raft_loop());
-		}
 	} else {
 		was_seen = bit_clear(&raft->leader_witness_map, source);
 		if (was_seen)
@@ -1086,6 +1094,34 @@ raft_notify_is_leader_seen(struct raft *raft, bool is_leader_seen,
 	return was_seen != is_leader_seen;
 }
 
+static bool
+raft_leader_see(struct raft *raft)
+{
+	uint32_t source = raft->self;
+	assert(source > 0 && source < VCLOCK_MAX);
+	bool was_seen = bit_set(&raft->leader_witness_map, source);
+	raft->leader_last_seen = raft_ev_monotonic_now(raft_loop());
+	if (!was_seen) {
+		raft_schedule_broadcast(raft);
+		return true;
+	}
+	return false;
+}
+
+static bool
+raft_leader_unsee(struct raft *raft)
+{
+	uint32_t source = raft->self;
+	assert(source > 0 && source < VCLOCK_MAX);
+	bool was_seen = bit_clear(&raft->leader_witness_map, source);
+	if (was_seen) {
+		raft_sm_election_update(raft);
+		raft_schedule_broadcast(raft);
+		return true;
+	}
+	return false;
+}
+
 void
 raft_cfg_is_candidate_later(struct raft *raft, bool is_candidate)
 {

```
</details>

Reply:
It might be enough to make it re-enterable until a next yield. I.e. check for is_write_in_progress only. Then first call will start election, next calls during the same event loop iteration won't do anything. And that would allow to avoid the ugly `if` in `raft_sm_election_update_cb()`. As an alternative to inlining `raft_leader_unsee()` call in this callback as I proposed in the new review.

Reply:
I think it's better to inline the `raft_leader_unsee`. Thanks for the suggestions!